bokehjs/src/lib/models/formatters/datetime_tick_formatter.ts

```diff
@@ -20,48 +18,62 @@ function _array(t: number): number[] {
   return tz(t, "%Y %m %d %H %M %S").split(/\s+/).map(e => parseInt(e, 10))
 }
 
-function _strftime(t: number, format: string | ((t: number) => string)): string {
-  if (isFunction(format)) {
-    return format(t)
-  } else {
-    // Python's datetime library augments the microsecond directive %f, which is not
-    // supported by the javascript library timezone: http://bigeasy.github.io/timezone/.
-    // Use a regular expression to replace %f directive with microseconds.
-    // TODO: what should we do for negative microsecond strings?
-    const microsecond_replacement_string = sprintf("$1%06d", _us(t))
-    format = format.replace(/((^|[^%])(%%)*)%f/, microsecond_replacement_string)
-
-    if (format.indexOf("%") == -1) {
-      // timezone seems to ignore any strings without any formatting directives,
-      // and just return the time argument back instead of the string argument.
-      // But we want the string argument, in case a user supplies a format string
-      // which doesn't contain a formatting directive or is only using %f.
-      return format
-    }
-
-    return tz(t, format)
+function _strftime(t: number, format: string): string {
+  // Python's datetime library augments the microsecond directive %f, which is not
+  // supported by the javascript library timezone: http://bigeasy.github.io/timezone/.
+  // Use a regular expression to replace %f directive with microseconds.
+  // TODO: what should we do for negative microsecond strings?
+  const microsecond_replacement_string = sprintf("$1%06d", _us(t))
+  format = format.replace(/((^|[^%])(%%)*)%f/, microsecond_replacement_string)
+
+  if (format.indexOf("%") == -1) {
+    // timezone seems to ignore any strings without any formatting directives,
+    // and just return the time argument back instead of the string argument.
+    // But we want the string argument, in case a user supplies a format string
+    // which doesn't contain a formatting directive or is only using %f.
+    return format
   }
+
+  return tz(t, format)
 }
 
+type formatType = "microseconds" | "milliseconds" | "seconds" | "minsec" | "minutes" | "hourmin" | "hours" | "days" | "months" | "years"
+
 // Labels of time units, from finest to coarsest.
-const format_order = [
+const format_order: formatType[] = [
   "microseconds", "milliseconds", "seconds", "minsec", "minutes", "hourmin", "hours", "days", "months", "years",
 ]
 
+// This dictionary maps the name of a time resolution (in @format_order)
+// to its index in a time.localtime() timetuple.  The default is to map
+// everything to index 0, which is year.  This is not ideal; it might cause
+// a problem with the tick at midnight, january 1st, 0 a.d. being incorrectly
+// promoted at certain tick resolutions.
+const time_tuple_ndx_for_resol: {[key: string]: number} = {}
```

Comment:
```suggestion
const time_tuple_ndx_for_resol: {[key in FormatType]: number} = {}
```
Same later in the code. Perhaps it would be even better to start using `Map<FormatType, number>` as it doesn't give room for misinterpretation of keys.

Reply:
https://github.com/bokeh/bokeh/pull/12170/commits/1525b8971bbc842ac377c87411970dbd7aadc2b5